fix(ngcc): correctly detect imported TypeScript helpers

gkalpak opened the pull request:
The `NgccReflectionHost`'s have logic for detecting certain known declarations (such as `Object.assign()` and TypeScript helpers), which allows the `PartialEvaluator` to evaluate expressions it would not be able to evaluate otherwise.

In #36089, `DelegatingReflectionHost` was introduced, which delegates to a TypeScript `ReflectionptHost` when reflecting on TypeScript files, which for ngcc's case means `.d.ts` files of dependencies. As a result, ngcc lost the ability to detect imported TypeScript helpers from `tslib`, because `DelegatingReflectionHost` was not able to apply the known declaration detection logic while reflecting on `tslib`'s `.d.ts` files.

This commit fixes this by ensuring `DelegatingReflectionHost` calls the `NgccReflectionHost`'s `addKnownDeclaration()` method as necessary, even when using the TypeScript `ReflectionHost`.

##
This is needed for angular/ngcc-validation#1029.

mary-poppins commented:
You can preview 6a34b64 at https://pr36284-6a34b64.ngbuilds.io/.

mary-poppins commented:
You can preview d9d8ffc at https://pr36284-d9d8ffc.ngbuilds.io/.

mary-poppins commented:
You can preview 415b579 at https://pr36284-415b579.ngbuilds.io/.

mary-poppins commented:
You can preview 0ae0ab3 at https://pr36284-0ae0ab3.ngbuilds.io/.

mary-poppins commented:
You can preview 58fe746 at https://pr36284-58fe746.ngbuilds.io/.

JoostK commented on packages/compiler-cli/ngcc/test/host/commonjs_host_spec.ts:
```diff
@@ -45,6 +47,12 @@ runInEachFileSystem(() => {
     let MODULE_WITH_PROVIDERS_PROGRAM: TestFile[];
 
+    // Helpers
+    const createHost = (bundle: BundleProgram, ngccHost: CommonJsReflectionHost) => {
```
Interestingly, I did this exact same thing when starting work on #36089 but it didn't end up in the end result. My reasoning for not going with it originally is that these tests are specific the the bundle format, not any .d.ts interop. However, this PR shows that there _is_ some sort of an interaction between the two, in which case I agree that changing the tests like this is the best way to go.

JoostK commented on packages/compiler-cli/ngcc/src/host/ngcc_host.ts:
```diff
@@ -128,4 +128,13 @@ export interface NgccReflectionHost extends ReflectionHost {
+   * @param decl The `Declaration` to check or `null` if there is no declaration.
+   * @return The passed in `Declaration` (potentially enhanced with a `KnownDeclaration`).
+   */
+  addKnownDeclaration<T extends Declaration|null>(decl: T): T;
```
I am not sold on the `add` part of the name, as my association is that this would record and track the declaration somehow, whereas it's more about the _detection_ of known declarations. So I would suggest something along the line of `detectKnownDeclaration` or `augmentKnownDeclaration`.

JoostK reviewed (commented):
Thank you for fixing what I broke! The approach overall looks good, is there a way we can test the last commit in this PR? Or would the refactored tests start to fail without it? If that were the cases, I think we should mention this in the commit message.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -602,8 +592,37 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+   */
+  addKnownDeclaration<T extends Declaration|null>(decl: T): T {
+    if (decl !== null && decl.known === null &&
+        // For some reason, TS cannot infer that `decl` is not `null`chere :/
```
If you change the method declaration throughout to this:

``` 
addKnownDeclaration<T extends Declaration>(decl: T|null): T|null;
```

Then TS is able to infer the non-null

gkalpak replied:
I went with `addKnownDeclaration<T extends Declaration|null>(decl: T): T`, because I wanted TS to be able to infer that when `null` is passed in, `null` will be returned (vs `Declaration | null`).

I now realize your suggestion combined with type overloads will work:

```ts
addKnownDeclaration(decl: null): null;
addKnownDeclaration<T extends Declaration>(decl: T): T;
addKnownDeclaration<T extends Declaration>(decl: T | null): T | null { ... }
```

Thx!

petebacondarwin reviewed (dismissed):
I am also not keen on the naming of `addKnownDeclaration()` but I would go further and suggest that we make this lazy...  How about we have a method on NgccReflectionHost:

```
getKnownDeclaration(decl: Declaration): ResolvedValue|null;
```

Then in the ES2015 reflectionhost it would look like:

```
getKnownDeclaration(decl: Declaration): ResolvedValue|null {
  if(decl.known !== undefined) {
    return decl.known;
  }
  if (this.isJavaScriptObjectDeclaration(decl)) {
    decl.known = resolveKnownDeclaration(KnownDeclaration.JsGlobalObject);
  } else {
    decl.known = null;
  }
  return decl;
}
```

```
getKnownDeclaration(decl: Declaration): ResolvedValue|null {
  if(decl.known !== undefined) {
    return decl.known;
  }
  const known = super.getKnownDeclaration(decl);
  if (known !== null) {
    const helperDecl = getTsHelperFnFromDeclaration(decl.node);
    decl.known = helperDecl && resolveKnownDeclaration(helperDecl);
  }
  return decl.known;
}
```

Then wherever there is currently a `decl.known` property reference in this PR, we would just call `host.getKnownDeclaration(decl)`.

gkalpak committed:
refactor(ngcc): move logic for identifying known declarations to method

The `NgccReflectionHost`s have logic for detecting certain known
declarations (such as `Object.assign()` and TypeScript helpers), which
allows the `PartialEvaluator` to evaluate expressions it would not be
able to statically evaluate otherwise.

This commit moves the logic for identifying these known declarations to
dedicated methods. This is in preparation of allowing ngcc's
`DelegatingReflectionHost` (introduced in #36089) to also apply the
known declaration detection logic when reflecting on TypeScript sources.

gkalpak committed:
test(ngcc): use `DelegatingReflectionHost` for testing `NgccReflectionHost`s

In #36089, `DelegatingReflectionHost` was introduced. Under the hood, it
delegates another `NgccReflectionHost` in order to reflect over the
program's source files, while using a different TypeScript
`ReflectionHost` to reflect over `.d.ts` files (which is how external
dependencies are represented in the program).

Previously, the `NgccReflectionHost`s were used directly in tests. This
does not exercise them in the way they are exercised in the actual
program, because (when used directly) they will also reflect on `.d.ts`
files too (instead of delegating to the TypeScript `ReflectionHost`).
This could hide bugs that would happen on the actual program.

This commit fixes this by using the `DelegatingReflectionHost` in the
various `NgccReflectionHost` tests.

NOTE:
This change will cause some of the existing tests to start failing.
These failures demonstrate pre-existing bugs in ngcc, that were hidden
due to the tests' being inconsistent with how the `ReflectionHost`s are
used in the actual program. They will be fixed in the next commit.

gkalpak committed:
fix(ngcc): correctly detect imported TypeScript helpers

The `NgccReflectionHost`s have logic for detecting certain known
declarations (such as `Object.assign()` and TypeScript helpers), which
allows the `PartialEvaluator` to evaluate expressions it would not be
able to statically evaluate otherwise.

In #36089, `DelegatingReflectionHost` was introduced, which delegates to
a TypeScript `ReflectionHost` when reflecting on TypeScript files, which
for ngcc's case means `.d.ts` files of dependencies. As a result, ngcc
lost the ability to detect TypeScript helpers imported from `tslib`,
because `DelegatingReflectionHost` was not able to apply the known
declaration detection logic while reflecting on `tslib`'s `.d.ts` files.

This commit fixes this by ensuring `DelegatingReflectionHost` calls the
`NgccReflectionHost`'s `detectKnownDeclaration()` method as necessary,
even when using the TypeScript `ReflectionHost`.

NOTE:
The previous commit exposed a bug in ngcc that was hidden due to the
tests' being inconsistent with how the `ReflectionHost`s are used in the
actual program. The changes in this commit are verified by ensuring the
failing tests are now passing (hence no new tests are added).

gkalpak committed:
fixup! refactor(ngcc): move logic for identifying known declarations to method

gkalpak committed:
fixup! fix(ngcc): correctly detect imported TypeScript helpers

gkalpak commented:
@JoostK, regarding the tests, yes, the last commit fixes the failing tests that were exposed by the previous commit. I have updated the commit messages to explain that.

mary-poppins commented:
You can preview ba674f9 at https://pr36284-ba674f9.ngbuilds.io/.

JoostK reviewed (approved):
LGTM. There's few typos in the commit message of cacc0f532638480ca7cf1642b4b38547d3ec9909:

- Paragraph 2 refers to `ReflectionptHost`, instead of `ReflectionHost`.
- Paragraph 3 refers to `addKnownDeclaration` which was renamed
- Paragraph 4 contains `tests'`, which _I think_ should be without the apostrophe.

<sup>I wish GH had suggestions for commit messages!</sup>

mary-poppins commented:
You can preview cf7bb1f at https://pr36284-cf7bb1f.ngbuilds.io/.

gkalpak commented:
Thx for the thorough review, @JoostK :100:
I've fixed the typos (and a couple more I noticed - <sub>Note to self: RTFCM before you submit</sub>) except for the `tests'`, which I am pretty sure is also correct (even if an unusual form :grin:).
